controllers/user/user_controller.go

```diff
@@ -18,48 +18,59 @@ import (
 	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+// +kubebuilder:rbac:groups=user.openshift.io,resources=groups,verbs=get;list;watch;create
+// +kubebuilder:rbac:groups=user.openshift.io,resources=groups,resourceNames=rhmi-developers,verbs=update;delete
+// +kubebuilder:rbac:groups=user.openshift.io,resources=users,verbs=watch;get;list
+
 var log = l.NewLoggerWithContext(l.Fields{l.ControllerLogContext: "user_controller"})
 
 // UserReconciler reconciles a User object
 type UserReconciler struct {
+	k8sclient.Client
+	Scheme *runtime.Scheme
+	mgr    manager.Manager
 }
 
-// +kubebuilder:rbac:groups=user.openshift.io,resources=groups,verbs=get;list;watch;create
-// +kubebuilder:rbac:groups=user.openshift.io,resources=groups,resourceNames=rhmi-developers,verbs=update;delete
-// +kubebuilder:rbac:groups=user.openshift.io,resources=users,verbs=watch;get;list
-
-func (r *UserReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
-	log.Info("Reconciling User")
-
-	// new client to avoid caching issues
+func New(mgr manager.Manager) *UserReconciler {
 	restConfig := controllerruntime.GetConfigOrDie()
 	restConfig.Timeout = time.Second * 10
-	scheme := runtime.NewScheme()
-	err := rhmiv1alpha1.AddToSchemes.AddToScheme(scheme)
+
+	client, err := k8sclient.New(restConfig, k8sclient.Options{
+		Scheme: mgr.GetScheme(),
+	})
 	if err != nil {
-		return ctrl.Result{}, err
+		panic("could not setup k8s client for user controller")
```

Comment:
My understanding was a panic will crash the application, which in our case would cause a pod restart. Is this not what happens?

Reply:
Good question, i questioned it's use as well. On review the New Functions when they return an error that is followed by an os.exit, which is effectively the same. the container will bump and retry. Not returning the error here makes this more consistent with other uses of the New function and the resulting error handling is effectively the same result as returning.

Reply:
Ok that makes sense. I do think that calling os.exit is the cleaner way to exit. In this case i don't think it matters but if a panic is called will that still allow the calling of any defer methods?